### PR TITLE
herbstluftwm 0.8.3 -> 0.9.1 + add tests

### PR DESCRIPTION
**Pull request by ThibautMarty**
###### Motivation for this change

- update herbstluftwm to 0.9.0
- run the package's tests
- add a NixOS test (similar to i3wm test)

I checked that cross compilation still works (`crossSystem.config = "aarch64-unknown-linux-gnu"`).

Replace #87003.

###### Things done

- [x] Tested using sandboxing ([nix.useSandbox](https://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](https://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [x] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [x] Tested execution of all binary files (usually in `./result/bin/`)
- [x] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [x] Ensured that relevant documentation is up to date
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

### CODE REVIEW

File: pkgs/applications/window-managers/herbstluftwm/default.nix [hunk 1]
```diff
@@ -39,8 +40,29 @@ stdenv.mkDerivation rec {
...
+
   postPatch = ''
     patchShebangs doc/gendoc.py
+    substituteInPlace tests/conftest.py --replace "/usr/bin/env bash" ${runtimeShell}
+    substituteInPlace tests/test_herbstluftwm.py --replace "/usr/bin/env bash" ${runtimeShell}
+    substituteInPlace share/autostart --replace /etc/xdg/herbstluftwm $out/etc/xdg/herbstluftwm
+  '';
+
+  checkPhase = ''
+    python3 -m pytest ../tests
```
Comment (SuperSandro2000) [inline pkgs/applications/window-managers/herbstluftwm/default.nix:66]: Please use the ``python.interpreter``.

File: pkgs/applications/window-managers/herbstluftwm/default.nix [hunk 2]
```diff
@@ -38,10 +41,40 @@ stdenv.mkDerivation rec {
...
+  passthru = {
+    tests.herbstluftwm = nixosTests.herbstluftwm;
+  };
+
   meta = {
     description = "A manual tiling window manager for X";
     homepage = "https://herbstluftwm.org/";
     license = stdenv.lib.licenses.bsd2;
     platforms = stdenv.lib.platforms.linux;
+    maintainers = with stdenv.lib.maintainers; [ thibautmarty ];
```
Comment (SuperSandro2000) [inline pkgs/applications/window-managers/herbstluftwm/default.nix:78]:
```suggestion
  meta = with stdenv.lib; {
    description = "A manual tiling window manager for X";
    homepage = "https://herbstluftwm.org/";
    license = licenses.bsd2;
    platforms = platforms.linux;
    maintainers = with maintainers; [ thibautmarty ];
```

**Comment by SuperSandro2000** [comment]
Result of `nixpkgs-review pr 104920` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>1 package built:</summary>
  <ul>
    <li>herbstluftwm</li>
  </ul>
</details>

File: pkgs/applications/window-managers/herbstluftwm/default.nix (hunk 1 above)
Comment (ThibautMarty) [reply pkgs/applications/window-managers/herbstluftwm/default.nix]: I would need to use `(python3.withPackages (ps: with ps; [ ewmh pytest xlib ])).interpreter` to make it works (or wrap it with a `let … in`). Is that what you want?

**Comment by berbiche** [comment]
@ThibautMarty @SuperSandro2000 triage a rebuild? I would like to see this merged thanks!

File: nixos/tests/all-tests.nix [hunk 3]
```diff
@@ -142,6 +142,7 @@ in
   handbrake = handleTestOn ["x86_64-linux"] ./handbrake.nix {};
   haproxy = handleTest ./haproxy.nix {};
   hardened = handleTest ./hardened.nix {};
+  herbstluftwm = handleTest ./herbstluftwm.nix {}
```
Comment (SuperSandro2000) [inline nixos/tests/all-tests.nix:145]:
```suggestion
  herbstluftwm = handleTest ./herbstluftwm.nix {};
```

**Review by SuperSandro2000 — CHANGES_REQUESTED** [review]
Also please squash all herbstluftwm init related commits together.

**Comment by ThibautMarty** [comment]
> Also please squash all herbstluftwm init related commits together.

What are "herbstluftwm init" commits? This not an "init package" PR, herbstluftwm was already packaged.
Each commit corresponds to 1 and only 1 thing:
1. add maintainer
2. add maintainership to the package
3. update the package
4. enable the package's test
5. add a NixOS test for the module using the package.

I'd rather prefer to not merge (3) and (4) (or even (5)), the distinction between the two can be useful when bisecting/cherry-picking/retrieving patchs/etc. But I will if you prefer.

**Comment by ThibautMarty** [comment]
Converted to draft because 0.9.1 had been released. I will update the package in a few days.

File: pkgs/applications/window-managers/herbstluftwm/default.nix [hunk 4]
```diff
@@ -1,24 +1,26 @@
 { stdenv, fetchurl, cmake, pkgconfig, glib, libX11, libXext, libXinerama, libXrandr
-, withDoc ? stdenv.buildPlatform == stdenv.targetPlatform, asciidoc ? null }:
+, python3, xdotool, xorg, xterm, runtimeShell
+, withDoc ? stdenv.buildPlatform == stdenv.targetPlatform, asciidoc ? null
```
Comment (SuperSandro2000) [inline pkgs/applications/window-managers/herbstluftwm/default.nix:3]:
```suggestion
, withDoc ? stdenv.buildPlatform == stdenv.targetPlatform, asciidoc
```

File: pkgs/applications/window-managers/herbstluftwm/default.nix [hunk 5]
```diff
@@ -1,24 +1,26 @@
...
+, nixosTests }:
 
 # Doc generation is disabled by default when cross compiling because asciidoc
-# does not cross compile for now
+# dependency is broken when cross compiling for now
 
 assert withDoc -> asciidoc != null;
```
Comment (SuperSandro2000) [inline pkgs/applications/window-managers/herbstluftwm/default.nix:10]:
```suggestion
```

File: pkgs/applications/window-managers/herbstluftwm/default.nix [hunk 6]
```diff
@@ -36,12 +38,53 @@ stdenv.mkDerivation rec {
...
+    substituteInPlace tests/test_herbstluftwm.py --replace "/usr/bin/env bash" ${runtimeShell}
+  '';
+
+  doCheck = true;
+
+  checkInputs = [
+    (python3.withPackages (ps: with ps; [ ewmh pytest xlib ]))
+    xdotool
+    xorg.xorgserver
+    xorg.xsetroot
```
Comment (SuperSandro2000) [inline pkgs/applications/window-managers/herbstluftwm/default.nix:67]:
```suggestion
    xorgserver
    xsetroot
```

File: pkgs/applications/window-managers/herbstluftwm/default.nix [hunk 7]
```diff
@@ -1,24 +1,26 @@
 { stdenv, fetchurl, cmake, pkgconfig, glib, libX11, libXext, libXinerama, libXrandr
-, withDoc ? stdenv.buildPlatform == stdenv.targetPlatform, asciidoc ? null }:
+, python3, xdotool, xorg, xterm, runtimeShell
```
Comment (SuperSandro2000) [inline pkgs/applications/window-managers/herbstluftwm/default.nix:2]:
```suggestion
, python3, xdotool, xorgserver, xsetroot, xterm, runtimeShell
```

File: pkgs/applications/window-managers/herbstluftwm/default.nix (hunk 4 above)
Comment (ThibautMarty) [reply pkgs/applications/window-managers/herbstluftwm/default.nix:3]: From what I see in nixpkgs, optional dependencies (enabled with withX, enableX, Xsupport, etc) are defaulted to null to avoid throwing an error and an assertion is added. See [pkgs/applications/video/mplayer/default.nix](https://github.com/NixOS/nixpkgs/blob/4b610d814e09857c974ac2f65374a5baba0902ff/pkgs/applications/video/mplayer/default.nix) for instance. Unless this has changed recently?

File: pkgs/applications/window-managers/herbstluftwm/default.nix (hunk 7 above)
Comment (ThibautMarty) [reply pkgs/applications/window-managers/herbstluftwm/default.nix]: done

**Comment by ThibautMarty** [comment]
I changed `python3` from buildInputs to nativeBuildInputs as it is only used at build-time for the documentation generation.
I did not notice this earlier because documentation generation is disabled for now (because of asciidoc).

Please see my comment in the review for "asciidoc ? null" and the assertion.

File: pkgs/applications/window-managers/herbstluftwm/default.nix (hunk 4 above)
Comment (SuperSandro2000) [reply pkgs/applications/window-managers/herbstluftwm/default.nix:3]: This is just syntactic sugar and is not required if you use lib.optionals.

Comment (ThibautMarty) [reply pkgs/applications/window-managers/herbstluftwm/default.nix:3]: No, it is not syntactic sugar. The semantic differs. `asciidoc ? null` ensures the derivation can be evaluated *even* if asciidoc has not been found by `callPackage`. When asciidoc is not needed we must not enforce this false dependency. Thus the common pattern is :
- defaulting asciidoc to `null`, so `callPackage` may not find it : this is not an error.
- ensuring we actually have asciidoc *only* when needed, with an assertion.
(I agree that, in this specific case and *for now*, asciidoc is overridden in all-packages.nix anyway.)
This is a common pattern in nixpkgs and it makes sense.

Comment (SuperSandro2000) [reply pkgs/applications/window-managers/herbstluftwm/default.nix:3]: > No, it is not syntactic sugar.

It is. Not required. At all. See how frameworks are handled on darwin. Nothing of the frameworks evals on linux but trough the usage of lib.opional(s) we do not evaluate it.

> asciidoc ? null ensures the derivation can be evaluated even if asciidoc has not been found by callPackage

> defaulting asciidoc to null, so callPackage may not find it : this is not an error.

That is wrong. The variable only needs to be defined. It can even be defined to null and evaluating won't fail.

I even gave this a try but a dependency wouldn't build with it but it evaluated.

> When asciidoc is not needed we must not enforce this false dependency.

This is only if doc is build. Otherwise nothing happens. asciidoc is only build and used by your package if withDoc is true. If it is nill or false and I removed all your extra code it is still not build for asciidoc.

> ensuring we actually have asciidoc only when needed, with an assertion.

That is not the case. Right now you are checking asciidoc is not overwritten with null.

If it would be null than you couldn't build herbstluftwm at all because ninja would fail to build. This proves you did not properly test this flag.

> I agree that, in this specific case and for now, asciidoc is overridden in all-packages.nix anyway.

It is not overwritten but the inputs is assigned a different version. It does not influence any of the things in here and makes no difference.

> This is a common pattern in nixpkgs and it makes sense.

It is a relic of the past and in most places we could remove it without any impact.

To prove my point I build this PR with the below diff without any eval errors:

```diff
diff --git a/pkgs/applications/window-managers/herbstluftwm/default.nix b/pkgs/applications/window-managers/herbstluftwm/default.nix
index 17a9784a339..ab5280d9686 100644
--- a/pkgs/applications/window-managers/herbstluftwm/default.nix
+++ b/pkgs/applications/window-managers/herbstluftwm/default.nix
@@ -1,14 +1,10 @@
 { stdenv, fetchurl, cmake, pkgconfig, python3, glib, libX11, libXext, libXinerama, libXrandr
 , xdotool, xorgserver, xsetroot, xterm, runtimeShell
-, withDoc ? stdenv.buildPlatform == stdenv.targetPlatform, asciidoc ? null
+, withDoc ? false
 , nixosTests }:
 
-# Doc generation is disabled by default when cross compiling because asciidoc
-# dependency is broken when cross compiling for now
-
-assert withDoc -> asciidoc != null;
-
 stdenv.mkDerivation rec {
+  asciidoc = null;
   pname = "herbstluftwm";
   version = "0.9.1";
 
diff --git a/pkgs/top-level/all-packages.nix b/pkgs/top-level/all-packages.nix
index fa9f0fe1de5..1eb4686fae1 100644
--- a/pkgs/top-level/all-packages.nix
+++ b/pkgs/top-level/all-packages.nix
@@ -22164,7 +22164,6 @@ in
   herbe = callPackage ../applications/misc/herbe { };
 
   herbstluftwm = callPackage ../applications/window-managers/herbstluftwm {
-    asciidoc = asciidoc-full;
   };
 
   hercules = callPackage ../applications/virtualization/hercules { };
```


But it failed to build because the the doc tests ``../tests/test_doc.py`` failed.

So please accept my suggestions and fix the flag you wanted to introduce because right now it is broken.

Comment (ThibautMarty) [reply pkgs/applications/window-managers/herbstluftwm/default.nix:3]: I think you missed my point: this is about the `callPackage` mechanism. *Not the evaluation nor the realisation*. Consider this:
```
> callPackage ({ doesNotExist, hello }: hello) {}
error: anonymous function at (string):1:14 called without required argument 'doesNotExist'
```
You'll get this error while `doesNotExist` *is* a false dependency. So things *happens*.

I will not answer all of your comments (but I can if you want) because I think we are both wasting time since we did not discussed the same thing. You try to demonstrate something which is not about what I talked about. (Some sentences show that you clearly did not understand what I mean – I apologize if I was not clear enough. Your diff removes the assertion that asciidoc is not null, set it to null and you say that the tests fail. Obviously. But what are you trying to prove?)

That being said, I'm willing to do the change (as I said "unless this has changed recently?") if this is the current pattern in nixpkgs (even though I think that's not a good thing and is a loss of information). Could you just direct me to a discussion/PR/issue/comment/RFC stating that? Thanks.

Comment (SuperSandro2000) [reply pkgs/applications/window-managers/herbstluftwm/default.nix:3]: > But what are you trying to prove?

That it is not required and your assertion does nothing and we can save those characters.

Also you option is broken and does not work despite you having extra code which does nothing.

> You'll get this error while doesNotExist is a false dependency. So things happens.

If asciidoc is not defined ninja will not build which is required in the dependency tree of this one.

> not a good thing and is a loss of information

Having an option in inputs and the lib.optional is enough information for everyone to know that those packages are required if you want to enable that option.

> Could you just direct me to a discussion/PR/issue/comment/RFC stating that?

The github search is terrible and I will never find anything relevant in there. Assertion in packages are mostly useful when you have multiple options which can't be activated at once.

Packages that do it like this are mostly old and arguing that a lot of packages do it will never yield any change in this regard.

**Comment by musteresel** [comment]
Could we, like, get this merged?

**Comment by SuperSandro2000** [comment]
> Could we, like, get this merged?

No, because ``nix-build -E "with import ./. {}; herbstluftwm.override { withDoc = false; }"`` does not build which it should do.

**Comment by ThibautMarty** [comment]
I'll fix that. We can wait #109455 to be merged to rebase onto.

**Comment by ThibautMarty** [comment]
Done. I removed the `withDoc` input as my initial intent was just disabling documentation generation when cross compiling because asciidoc package does not work well with cross compilation and I'm tired of explaining my choices. So let's just remove this from the user perspective as I'm sure nobody never bothered using it explicitly. This removes the broken combination : running tests + no documentation generation so everything is good.
The package builds both for native and cross compilation, package's tests pass, I added a NixOS test. Hopefully we can get back to trivial automatic updates after this, finally.

**Comment by SuperSandro2000** [comment]
This is a semi-automatic executed nixpkgs-review which is checked by a human on a best effort basis and does not build all packages (e.g. lumo, tensorflow or pytorch).
If you have any questions or problems please reach out to SuperSandro2000 on IRC.

Result of `nixpkgs-review pr 104920` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
<summary>
1 package built:
</summary>
<ul>
<li>herbstluftwm</li>
</ul>
</details>
